File: Php/framework/framework/smartmoney.php
<?php

namespace Smartkernel\Framework {

    use Smartkernel\Framework\SmartString;
    use Smartkernel\Framework\SmartMath;

/*
     * 金钱
     */

    class SmartMoney {

        private static function splitToChinese($input, $index) {
            $result = array('零', '壹', '贰', '叁', '肆', '伍', '陆', '染', '捌', '玖');
            $length = count($input);
            if ($index > $length - 1) {
                return "零";
            }
            return $result[intval($input[$length - 1 - $index])];
        }

        /*
         * 金额转换为中文大写形式
         * 
         * @param float $money，金额
         * @param bool $isFixLength，是否固定长度，如果为false，将删除无效的零
         * @return string，中文大写形式
         * 
         */

        public static function toChinese($money, $isFixLength = true) {
            $result = "{17}仟{16}佰{15}拾{14}万{13}仟{12}佰{11}拾{10}億{9}仟{8}佰{7}拾{6}萬{5}仟{4}佰{3}拾{2}元{1}角{0}分";
            $format = SmartString::toCharArray(SmartString::replace(sprintf("%.2f", $money), ".", ""));
            $result = SmartString::format($result, self::splitToChinese($format, 0), self::splitToChinese($format, 1), self::splitToChinese($format, 2), self::splitToChinese($format, 3), self::splitToChinese($format, 4), self::splitToChinese($format, 5), self::splitToChinese($format, 6), self::splitToChinese($format, 7), self::splitToChinese($format, 8), self::splitToChinese($format, 9), self::splitToChinese($format, 10), self::splitToChinese($format, 11), self::splitToChinese($format, 12), self::splitToChinese($format, 13), self::splitToChinese($format, 14), self::splitToChinese($format, 15), self::splitToChinese($format, 16), self::splitToChinese($format, 17));
            if (!$isFixLength) {
                $result = SmartString::replaceMore($result, array("零角", "零分"), "");
                $result = SmartString::replaceMore($result, array("零拾", "零佰", "零仟", "零万"), "零");
                $result = SmartString::replace($result, "零萬", "萬");
                $result = SmartString::replace($result, "零億", "億");
                $result = SmartString::repeatReplace($result, "零元", "元");
                $result = SmartString::repeatReplace($result, "零萬", "萬");
                $result = SmartString::repeatReplace($result, "零億", "億");
                $result = SmartString::repeatReplace($result, "零零", "零");
                $result = SmartString::replace($result, "億萬", "億零");
                $result = SmartString::replace(SmartString::trimStart($result, '億'), "万", "萬");
                $result = SmartString::repeatTrimStart($result, "零");
                return $result;
            }
            return SmartString::replace($result, "万", "萬");
        }

        /*
         * 数值转换为英文形式
         * 
         * @param float $money，数量
         * @return string，英文形式
         * 
         */

        public static function toEnglish($money) {
            return (new SmartEnglishConverter())->convert($money);
        }

    }

    /*
     * 英文转换器
     */

    class SmartEnglishConverter {

        private $figs = array();
        private $tens = array();
        private $units = array();

        /*
         * 转换
         * 
         * @param float $number，数字
         * @return string，转换的结果
         * 
         */

        public function convert($number) {
            $left = "";
            $right = "";
            $temp = "";
            self::init();
            $result = strval(SmartMath::round($number, 2));
            if (SmartString::indexOf($result, ".") == -1) {
                $left = $result;
                $right = "";
            } else {
                $left = SmartString::substring($result, 0, SmartString::indexOf($result, "."));
                $right = SmartString::substring($result, SmartString::indexOf($result, ".") + 1, SmartString::length($result) - SmartString::indexOf($result, ".") - 1);
            }
            if (SmartString::length($left) > 12) {
                return null;
            }
            $result = "";
            while (SmartString::length($left) > 0) {
                $length = self::len($left);
                $current = "";
                if ($length % 3 == 0) {
                    $current = self::left($left, 3);
                    $left = self::right($left, $length - 3);
                } else {
                    $current = self::left($left, ($length % 3));
                    $left = self::right($left, $length - ($length % 3));
                }
                $nbit = self::len($left) / 3;
                $temp = self::decodeHundred($current);
                if (($left == strval(self::len($left)) || $nbit == 0) && self::len($current) == 3) {
                    if (intval(self::left($current, 1)) != 0 & intval(self::right($current, 2)) != 0) {
                        $temp = self::left($temp, SmartString::indexOf($temp, $this->units[3]) + self::len($this->units[3])) . $this->units[7] . " " . self::right($temp, self::len($temp) - (SmartString::indexOf($temp, $this->units[3]) + self::len($this->units[3])));
                    } else {
                        $temp = $this->units[7] . " " . $temp;
                    }
                }
                if ($nbit == 0) {
                    $result = SmartString::trim(strval($result . " " . $temp));
                } else {
                    $result = SmartString::trim(strval($result . " " . $temp . " " . $this->units[$nbit - 1]));
                }
                if (self::left($result, 3) == $this->units[7]) {
                    $result = SmartString::trim(strval(self::right($result, self::len($result) - 3)));
                }
                if ($left == strval(self::len($left))) {
                    return "";
                }
            }
            $left = $result;
            if (self::len($right) > 0) {
                $right = $this->units[5] . " " . self::decodeHundred($right) . " " . $this->units[6];
            } else {
                $right = $this->units[4];
            }
            return SmartString::toUpper(($left . " " . $right));
        }

        private function init() {
            $this->figs[0] = "One";
            $this->figs[1] = "Two";
            $this->figs[2] = "Three";
            $this->figs[3] = "Four";
            $this->figs[4] = "Five";
            $this->figs[5] = "Six";
            $this->figs[6] = "Seven";
            $this->figs[7] = "Eight";
            $this->figs[8] = "Nine";
            $this->figs[9] = "Ten";
            $this->figs[10] = "Eleven";
            $this->figs[11] = "Twelve";
            $this->figs[12] = "Thirteen";
            $this->figs[13] = "Fourteen";
            $this->figs[14] = "Fifteen";
            $this->figs[15] = "Sixteen";
            $this->figs[16] = "Seventeen";
            $this->figs[17] = "Eighteen";
            $this->figs[18] = "Nineteen";
            $this->tens[0] = "Ten";
            $this->tens[1] = "Twenty";
            $this->tens[2] = "Thirty";
            $this->tens[3] = "Forty";
            $this->tens[4] = "Fifty";
            $this->tens[5] = "Sixty";
            $this->tens[6] = "Seventy";
            $this->tens[7] = "Eighty";
            $this->tens[8] = "Ninety";
            $this->units[0] = "Thousand";
            $this->units[1] = "Million";
            $this->units[2] = "Billion";
            $this->units[3] = "Hundred";
            $this->units[4] = "Only";
            $this->units[5] = "Dollars and";
            $this->units[6] = "Cent";
            $this->units[7] = "";
        }

        private function decodeHundred($hundredString) {
            $rtn = "";
            if (self::len($hundredString) > 0 && self::len($hundredString) <= 3) {
                $tmp = 0;
                switch (self::len($hundredString)) {
                    case 1:
                        $tmp = intval($hundredString);
                        if ($tmp != 0) {
                            $rtn = $this->figs[$tmp - 1];
                        }
                        break;
                    case 2:
                        $tmp = intval($hundredString);
                        if ($tmp != 0) {
                            if (($tmp < 20)) {
                                $rtn = $this->figs[$tmp - 1];
                            } else {
                                $rtn = intval(self::right($hundredString, 1)) == 0 ? $this->tens[intval($tmp / 10) - 1] : strval($this->tens[intval($tmp / 10) - 1] . " " . $this->figs[intval(self::right($hundredString, 1)) - 1]);
                            }
                        }
                        break;
                    case 3:
                        $rtn = intval(self::left($hundredString, 1)) != 0 ? strval($this->figs[intval(self::left($hundredString, 1)) - 1] . " " . $this->units[3] . "AND " . self::decodeHundred(self::right($hundredString, 2))) : self::decodeHundred(self::right($hundredString, 2));
                        break;
                    default:
                        break;
                }
            }
            return $rtn;
        }

        private static function left($str, $n) {
            return SmartString::substring($str, 0, $n);
        }

        private static function right($str, $n) {
            return SmartString::substring($str, SmartString::length($str) - $n, $n);
        }

        private static function len($str) {
            return SmartString::length($str);
        }

    }

}

File: Framework/CSharp/Framework/Framework/SmartMoney.cs
/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */
using System;
using System.Collections.Generic;
using System.Linq;

namespace Smartkernel.Framework
{
	/// <summary>
	/// 金钱
	/// </summary>
	public static class SmartMoney
	{
		private static readonly SmartEnglishConverter englishConverter = new SmartEnglishConverter();

		/// <summary>
		/// 金额转换为中文大写形式
		/// </summary>
		/// <param name="money">金额</param>
		/// <param name="isFixLength">是否固定长度，如果为false，将删除无效的零</param>
		/// <returns>中文大写形式</returns>
		public static string ToChinese(decimal money, bool isFixLength = true)
		{
			string result = "{17}仟{16}佰{15}拾{14}万{13}仟{12}佰{11}拾{10}億{9}仟{8}佰{7}拾{6}萬{5}仟{4}佰{3}拾{2}元{1}角{0}分";
			var format = money.ToString("#.00").Replace(".", "");
			result = string.Format(result, SplitToChinese(format, 0), SplitToChinese(format, 1), SplitToChinese(format, 2), SplitToChinese(format, 3), SplitToChinese(format, 4), SplitToChinese(format, 5), SplitToChinese(format, 6), SplitToChinese(format, 7), SplitToChinese(format, 8), SplitToChinese(format, 9), SplitToChinese(format, 10), SplitToChinese(format, 11), SplitToChinese(format, 12), SplitToChinese(format, 13), SplitToChinese(format, 14), SplitToChinese(format, 15), SplitToChinese(format, 16), SplitToChinese(format, 17));
			if (!isFixLength)
			{
				result = SmartString.Replace(result, new[] { "零角", "零分" }, "");
				result = SmartString.Replace(result, new[] { "零拾", "零佰", "零仟", "零万" }, "零").Replace("零萬", "萬").Replace("零億", "億");
				result = SmartString.RepeatReplace(result, "零元", "元");
				result = SmartString.RepeatReplace(result, "零萬", "萬");
				result = SmartString.RepeatReplace(result, "零億", "億");
				result = SmartString.RepeatReplace(result, "零零", "零");
				result = result.Replace("億萬", "億零");
				result = result.TrimStart('億').Replace("万", "萬");
				result = SmartString.RepeatTrimStart(result, "零");
				return result;
			}
			return result.Replace("万", "萬");
		}

		private static string SplitToChinese(string input, int index)
		{
			var result = new[] { '零', '壹', '贰', '叁', '肆', '伍', '陆', '染', '捌', '玖' };
			if (index > input.Length - 1)
			{
				return "零";
			}
			return result[Convert.ToInt32(input[input.Length - 1 - index].ToString())].ToString();
		}

		/// <summary>
		/// 数值转换为英文形式
		/// </summary>
		/// <param name="money">数量</param>
		/// <returns>英文形式</returns>
		public static string ToEnglish(decimal money)
		{
			return englishConverter.Convert(money);
		}
	}

	/// <summary>
	/// 英文转换器
	/// </summary>
	internal class SmartEnglishConverter
	{
		private readonly string[] figs = new string[19];

		private readonly string[] tens = new string[9];

		private readonly string[] units = new string[8];

		/// <summary>
		/// 转换
		/// </summary>
		/// <param name="number">数字</param>
		/// <returns>转换的结果</returns>
		public string Convert(decimal number)
		{
			string left;
			string right;
			string temp;
			Init();
			string result = System.Convert.ToString(Math.Round(number, 2));
			if (result.IndexOf(".") == -1)
			{
				left = result;
				right = "";
			}
			else
			{
				left = result.Substring(0, result.IndexOf("."));
				right = result.Substring(result.IndexOf(".") + 1, result.Length - result.IndexOf(".") - 1);
			}
			if (left.Length > 12)
			{
				return null;
			}
			result = "";
			while (left.Length > 0)
			{
				int length = Len(left);
				string current;
				if (length % 3 == 0)
				{
					current = Left(left, 3);
					left = Right(left, length - 3);
				}
				else
				{
					current = Left(left, (length % 3));
					left = Right(left, length - (length % 3));
				}
				int nbit = Len(left) / 3;
				temp = DecodeHundred(current);
				if ((left == Len(left).ToString() || nbit == 0) && Len(current) == 3)
				{
					if (System.Convert.ToInt32(Left(current, 1)) != 0 & System.Convert.ToInt32(Right(current, 2)) != 0)
					{
						temp = Left(temp, temp.IndexOf(units[3]) + Len(units[3])) + units[7] + " " + Right(temp, Len(temp) - (temp.IndexOf(units[3]) + Len(units[3])));
					}
					else
					{
						temp = units[7] + " " + temp;
					}
				}
				if (nbit == 0)
				{
					result = System.Convert.ToString(result + " " + temp).Trim();
				}
				else
				{
					result = System.Convert.ToString(result + " " + temp + " " + units[nbit - 1]).Trim();
				}
				if (Left(result, 3) == units[7])
				{
					result = System.Convert.ToString(Right(result, Len(result) - 3)).Trim();
				}
				if (left == Len(left).ToString())
				{
					return "";
				}
			}
			left = result;
			if (Len(right) > 0)
			{
				right = units[5] + " " + DecodeHundred(right) + " " + units[6];
			}
			else
			{
				right = units[4];
			}
			return (left + " " + right).ToUpper();
		}

		private void Init()
		{
			if (figs[0] != "One")
			{
				figs[0] = "One";
				figs[1] = "Two";
				figs[2] = "Three";
				figs[3] = "Four";
				figs[4] = "Five";
				figs[5] = "Six";
				figs[6] = "Seven";
				figs[7] = "Eight";
				figs[8] = "Nine";
				figs[9] = "Ten";
				figs[10] = "Eleven";
				figs[11] = "Twelve";
				figs[12] = "Thirteen";
				figs[13] = "Fourteen";
				figs[14] = "Fifteen";
				figs[15] = "Sixteen";
				figs[16] = "Seventeen";
				figs[17] = "Eighteen";
				figs[18] = "Nineteen";
				tens[0] = "Ten";
				tens[1] = "Twenty";
				tens[2] = "Thirty";
				tens[3] = "Forty";
				tens[4] = "Fifty";
				tens[5] = "Sixty";
				tens[6] = "Seventy";
				tens[7] = "Eighty";
				tens[8] = "Ninety";
				units[0] = "Thousand";
				units[1] = "Million";
				units[2] = "Billion";
				units[3] = "Hundred";
				units[4] = "Only";
				units[5] = "Dollars and";
				units[6] = "Cent";
				units[7] = "";
			}
		}

		private string DecodeHundred(string hundredString)
		{
			var rtn = "";
			if (Len(hundredString) > 0 && Len(hundredString) <= 3)
			{
				int tmp;
				switch (Len(hundredString))
				{
					case 1:
						tmp = System.Convert.ToInt32(hundredString);
						if (tmp != 0)
						{
							rtn = figs[tmp - 1];
						}
						break;
					case 2:
						tmp = System.Convert.ToInt32(hundredString);
						if (tmp != 0)
						{
							if ((tmp < 20))
							{
								rtn = figs[tmp - 1];
							}
							else
							{
								rtn = System.Convert.ToInt32(Right(hundredString, 1)) == 0 ? tens[System.Convert.ToInt32(tmp / 10) - 1] : System.Convert.ToString(tens[System.Convert.ToInt32(tmp / 10) - 1] + " " + figs[System.Convert.ToInt32(Right(hundredString, 1)) - 1]);
							}
						}
						break;
					case 3:
						rtn = System.Convert.ToInt32(Left(hundredString, 1)) != 0 ? System.Convert.ToString(figs[System.Convert.ToInt32(Left(hundredString, 1)) - 1] + " " + units[3] + "AND " + DecodeHundred(Right(hundredString, 2))) : DecodeHundred(Right(hundredString, 2));
						break;
					default:
						break;
				}
			}
			return rtn;
		}

		private static string Left(string str, int n)
		{
			return str.Substring(0, n);
		}

		private static string Right(string str, int n)
		{
			return str.Substring(str.Length - n, n);
		}

		private static int Len(string str)
		{
			return str.Length;
		}
	}
}
